{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Pipeline\n",
    "\n",
    "This jupyter notebook deals with data preprocessing for further analysis in the PyClone, SciClone, TitanCNA, FastClone and PyClone-VI libraries. A prepared .vcf file and data obtained from the variant calling process are required for launch. For the preparation of variants, it is recommended to use the CNVpytor library, which ensures that the resulting data is immediately suitable for preprocessing without further modifications. When using other libraries, it is necessary to edit the data into the format below.\n",
    "\n",
    "## Copy Calling\n",
    "\n",
    "For copy calling, we recommend using CNVpytor or the CNVnator library. Both libraries are suitable for preparing variants as they work identically and return the same formats. CNVpytor serves as an extension of the CNVnator library for compatibility with the Python programming language.\n",
    "\n",
    "Specific details on the use of this tool can be found in the file **copyCalling.pdf**\n",
    "\n",
    "\n",
    "\n",
    "### Format of variant calling results file (.tsv)\n",
    "\n",
    "If you decide to use another library for the preparation of variants, you must follow the following guidelines. In general, this file must be in .tsv (Tab separated values) format. Values can be separated by tabs or number of whitespaces. The program does not recognizes other types of separators and will ignore them. **Columns required:**\n",
    "\n",
    "- CNV type - \"deletion\" or \"duplication\"\n",
    "- CNV region - chr:start-end\n",
    "- CNV size - end-start\n",
    "- CNV level - read depth normalized to 1\n",
    "- e-val1 - e-value (p-value multiplied by genome size divided by bin size) calculated using t-test statistics between RD statistics in the region and global\n",
    "- e-val2 - e-value (p-value multiplied by genome size divided by bin size) from the probability of RD values within the region to be in the tails of a gaussian distribution of binned RD\n",
    "- e-val3 - same as e-val1 but for the middle of CNV\n",
    "- e-val4 - same as e-val2 but for the middle of CNV\n",
    "- q0 - fraction of reads mapped with q0 quality in call region\n",
    "- pN - fraction of reference genome gaps (Ns) in call region\n",
    "- dG - distance from closest large (>100bp) gap in reference genome\n",
    "\n",
    "\n",
    "Subsequently, this jupyter notebook runs all tools and saves the necessary results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import main\n",
    "from sklearn.pipeline import Pipeline \n",
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading of an input VCF file. Stored in a separate cell for easier data reloading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>DO52567</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39261</td>\n",
       "      <td>rs200677948</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>42.64</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-0.431;DB;DP=3...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:1,2:3:20:50,0,20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55164</td>\n",
       "      <td>rs3091274</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>53.32</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AF=1.00;AN=2;DB;DP=2;ExcessHet=0.0000;FS=...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:0,2:2:6:65,6,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55249</td>\n",
       "      <td>rs200769871</td>\n",
       "      <td>C</td>\n",
       "      <td>CTATGG</td>\n",
       "      <td>78.28</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AF=1.00;AN=2;DB;DP=2;ExcessHet=0.0000;FS=...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:0,2:2:6:90,6,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55299</td>\n",
       "      <td>rs10399749</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>51.64</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=0.000;DB;DP=3;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:1,2:3:25:59,0,25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84010</td>\n",
       "      <td>rs186443818</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>56.84</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AF=1.00;AN=2;DB;DP=5;ExcessHet=0.0000;FS=...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:0,4:4:9:70,9,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM    POS           ID REF     ALT   QUAL FILTER  \\\n",
       "0  chr1  39261  rs200677948   T       C  42.64      .   \n",
       "1  chr1  55164    rs3091274   C       A  53.32      .   \n",
       "2  chr1  55249  rs200769871   C  CTATGG  78.28      .   \n",
       "3  chr1  55299   rs10399749   C       T  51.64      .   \n",
       "4  chr1  84010  rs186443818   G       A  56.84      .   \n",
       "\n",
       "                                                INFO          FORMAT  \\\n",
       "0  AC=1;AF=0.500;AN=2;BaseQRankSum=-0.431;DB;DP=3...  GT:AD:DP:GQ:PL   \n",
       "1  AC=2;AF=1.00;AN=2;DB;DP=2;ExcessHet=0.0000;FS=...  GT:AD:DP:GQ:PL   \n",
       "2  AC=2;AF=1.00;AN=2;DB;DP=2;ExcessHet=0.0000;FS=...  GT:AD:DP:GQ:PL   \n",
       "3  AC=1;AF=0.500;AN=2;BaseQRankSum=0.000;DB;DP=3;...  GT:AD:DP:GQ:PL   \n",
       "4  AC=2;AF=1.00;AN=2;DB;DP=5;ExcessHet=0.0000;FS=...  GT:AD:DP:GQ:PL   \n",
       "\n",
       "                DO52567  \n",
       "0  0/1:1,2:3:20:50,0,20  \n",
       "1    1/1:0,2:2:6:65,6,0  \n",
       "2    1/1:0,2:2:6:90,6,0  \n",
       "3  0/1:1,2:3:25:59,0,25  \n",
       "4    1/1:0,4:4:9:70,9,0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading and making dataframe out of .vcf file\n",
    "main_vcf = main.load_vcf_file('./DO52567.vcf')\n",
    "main_vcf.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading of an result file from Copy Caller. Stored in a separate cell for easier data reloading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>method</th>\n",
       "      <th>CNV Type</th>\n",
       "      <th>chr</th>\n",
       "      <th>CNV Region Start</th>\n",
       "      <th>CNV Region End</th>\n",
       "      <th>CNV size</th>\n",
       "      <th>CNV level</th>\n",
       "      <th>e-val1</th>\n",
       "      <th>e-val2</th>\n",
       "      <th>e-val3</th>\n",
       "      <th>e-val4</th>\n",
       "      <th>q0</th>\n",
       "      <th>pN</th>\n",
       "      <th>dG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>file</td>\n",
       "      <td>rd_mean_shift</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>180001.0</td>\n",
       "      <td>220000.0</td>\n",
       "      <td>40000.0</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>3.970129e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.970129e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>file</td>\n",
       "      <td>rd_mean_shift</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>260001.0</td>\n",
       "      <td>320000.0</td>\n",
       "      <td>60000.0</td>\n",
       "      <td>0.142411</td>\n",
       "      <td>8.189110e+00</td>\n",
       "      <td>1.907243e-35</td>\n",
       "      <td>8.189110e+00</td>\n",
       "      <td>1.907243e-35</td>\n",
       "      <td>0.916184</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>file</td>\n",
       "      <td>rd_mean_shift</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>470001.0</td>\n",
       "      <td>530000.0</td>\n",
       "      <td>60000.0</td>\n",
       "      <td>0.117525</td>\n",
       "      <td>3.338000e+00</td>\n",
       "      <td>2.064495e-60</td>\n",
       "      <td>3.338000e+00</td>\n",
       "      <td>2.064495e-60</td>\n",
       "      <td>0.675772</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>file</td>\n",
       "      <td>rd_mean_shift</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>1620001.0</td>\n",
       "      <td>1690000.0</td>\n",
       "      <td>70000.0</td>\n",
       "      <td>0.703857</td>\n",
       "      <td>3.192341e-01</td>\n",
       "      <td>1.552946e-09</td>\n",
       "      <td>3.192341e-01</td>\n",
       "      <td>1.552946e-09</td>\n",
       "      <td>0.372246</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>944400.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>file</td>\n",
       "      <td>rd_mean_shift</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>2630001.0</td>\n",
       "      <td>2680000.0</td>\n",
       "      <td>50000.0</td>\n",
       "      <td>0.084009</td>\n",
       "      <td>7.421955e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>7.421955e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.324025</td>\n",
       "      <td>0.915600</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  file_name         method  CNV Type chr  CNV Region Start  CNV Region End  \\\n",
       "0      file  rd_mean_shift  deletion   1          180001.0        220000.0   \n",
       "1      file  rd_mean_shift  deletion   1          260001.0        320000.0   \n",
       "2      file  rd_mean_shift  deletion   1          470001.0        530000.0   \n",
       "3      file  rd_mean_shift  deletion   1         1620001.0       1690000.0   \n",
       "4      file  rd_mean_shift  deletion   1         2630001.0       2680000.0   \n",
       "\n",
       "   CNV size  CNV level        e-val1        e-val2        e-val3  \\\n",
       "0   40000.0   0.000405  3.970129e-07  0.000000e+00  3.970129e-07   \n",
       "1   60000.0   0.142411  8.189110e+00  1.907243e-35  8.189110e+00   \n",
       "2   60000.0   0.117525  3.338000e+00  2.064495e-60  3.338000e+00   \n",
       "3   70000.0   0.703857  3.192341e-01  1.552946e-09  3.192341e-01   \n",
       "4   50000.0   0.084009  7.421955e+00  0.000000e+00  7.421955e+00   \n",
       "\n",
       "         e-val4        q0        pN        dG  \n",
       "0  0.000000e+00  1.000000  1.000000       0.0  \n",
       "1  1.907243e-35  0.916184  0.833333       0.0  \n",
       "2  2.064495e-60  0.675772  0.833333       0.0  \n",
       "3  1.552946e-09  0.372246  0.000000  944400.0  \n",
       "4  0.000000e+00  0.324025  0.915600       0.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading and making dataframe out of Copy Caller results file\n",
    "cnvnator_df = pd.read_csv('./copyCaller/results/output.tsv',  names=['file_name', 'method', 'CNV Type', 'chr', 'CNV Region Start', 'CNV Region End', 'CNV size', 'CNV level', 'e-val1', 'e-val2', 'e-val3', 'e-val4', 'q0', 'pN', 'dG'], delimiter=r\"\\s+\")\n",
    "cnvnator_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Main pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "There were 12 warnings (use warnings() to see them)\n",
      "Running TITAN...\n",
      "titan: Loading data ./InputFiles/TitanCNA/Titancna.het\n",
      "titan: Loading GC content and mappability corrected log2 ratios...\n",
      "titan: Extracting read depth...\n",
      "Removed Chrs: \n",
      "titan: Loading default parameters\n",
      "titan: Using 1 cores.\n",
      "titan: Parameter estimation\n",
      "Optimal state path computation: Using 1 cores.\n",
      "Writing results to ./Results/TitanCNA//TitanCNA_cluster01.titan.txt, ./Results/TitanCNA//TitanCNA_cluster01.segs.txt, ./Results/TitanCNA//TitanCNA_cluster01.params.txt\n",
      "titan: Saving parameters to ./Results/TitanCNA//TitanCNA_cluster01.params.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$dens.bw\n",
      "[1] 0.2387662\n",
      "\n",
      "$scat\n",
      "[1] 0.5538911\n",
      "\n",
      "$S_Dbw\n",
      "[1] 0.7926573\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fontconfig warning: ignoring UTF-8: not a valid region tag\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n",
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>DO52567</th>\n",
       "      <th>GENOTYPE</th>\n",
       "      <th>ALLELIC DEPTH</th>\n",
       "      <th>DEPTH</th>\n",
       "      <th>GENOTYPE QUALITY</th>\n",
       "      <th>PHRED-SCALED LIKELIHOODS</th>\n",
       "      <th>ALLELIC FREQUENCY</th>\n",
       "      <th>MINOR ALLELE COPY NUMBER</th>\n",
       "      <th>MAJOR ALLELE COPY NUMBER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>790733</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CAATGG</td>\n",
       "      <td>142.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-1.156;DP=14;E...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:6,4:10:99:150,0,219</td>\n",
       "      <td>0/1</td>\n",
       "      <td>6,4</td>\n",
       "      <td>10</td>\n",
       "      <td>99</td>\n",
       "      <td>150,0,219</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>chr1</td>\n",
       "      <td>855373</td>\n",
       "      <td>rs74609473</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "      <td>109.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-1.440;DB;DP=1...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:3,3:6:99:117,0,117</td>\n",
       "      <td>0/1</td>\n",
       "      <td>3,3</td>\n",
       "      <td>6</td>\n",
       "      <td>99</td>\n",
       "      <td>117,0,117</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1366183</td>\n",
       "      <td>rs199546086</td>\n",
       "      <td>TGA</td>\n",
       "      <td>T</td>\n",
       "      <td>104.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=0.598;DB;DP=7;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:3,3:6:99:112,0,112</td>\n",
       "      <td>0/1</td>\n",
       "      <td>3,3</td>\n",
       "      <td>6</td>\n",
       "      <td>99</td>\n",
       "      <td>112,0,112</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>669</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1898612</td>\n",
       "      <td>rs142234461</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "      <td>136.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-3.087;DB;DP=1...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:8,4:12:99:144,0,324</td>\n",
       "      <td>0/1</td>\n",
       "      <td>8,4</td>\n",
       "      <td>12</td>\n",
       "      <td>99</td>\n",
       "      <td>144,0,324</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1938054</td>\n",
       "      <td>rs113907178;rs35385292</td>\n",
       "      <td>ACT</td>\n",
       "      <td>A</td>\n",
       "      <td>148.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-0.180;DB;DP=8...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:4,4:8:99:156,0,156</td>\n",
       "      <td>0/1</td>\n",
       "      <td>4,4</td>\n",
       "      <td>8</td>\n",
       "      <td>99</td>\n",
       "      <td>156,0,156</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554666</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56881158</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>129.97</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-1.006;DP=13;E...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:4,4:8:99:139,0,99</td>\n",
       "      <td>0/1</td>\n",
       "      <td>4,4</td>\n",
       "      <td>8</td>\n",
       "      <td>99</td>\n",
       "      <td>139,0,99</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554696</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56885195</td>\n",
       "      <td>rs56282167</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>97.60</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=0.160;DB;DP=11...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:7,4:11:99:105,0,225</td>\n",
       "      <td>0/1</td>\n",
       "      <td>7,4</td>\n",
       "      <td>11</td>\n",
       "      <td>99</td>\n",
       "      <td>105,0,225</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554697</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56885248</td>\n",
       "      <td>rs55700828</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>109.64</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=-0.815;DB;DP=1...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,5:15:99:117,0,272</td>\n",
       "      <td>0/1</td>\n",
       "      <td>10,5</td>\n",
       "      <td>15</td>\n",
       "      <td>99</td>\n",
       "      <td>117,0,272</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554706</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886312</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>89.64</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=1.253;DP=13;Ex...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:8,4:12:97:97,0,197</td>\n",
       "      <td>0/1</td>\n",
       "      <td>8,4</td>\n",
       "      <td>12</td>\n",
       "      <td>97</td>\n",
       "      <td>97,0,197</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554707</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886334</td>\n",
       "      <td>rs4047361</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>109.64</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AF=0.500;AN=2;BaseQRankSum=0.256;DB;DP=14...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:7,5:12:99:117,0,183</td>\n",
       "      <td>0/1</td>\n",
       "      <td>7,5</td>\n",
       "      <td>12</td>\n",
       "      <td>99</td>\n",
       "      <td>117,0,183</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36927 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM       POS                      ID  REF     ALT    QUAL FILTER  \\\n",
       "42       chr1    790733                       .    C  CAATGG  142.60      .   \n",
       "91       chr1    855373              rs74609473  CAT       C  109.60      .   \n",
       "441      chr1   1366183             rs199546086  TGA       T  104.60      .   \n",
       "669      chr1   1898612             rs142234461  CAT       C  136.60      .   \n",
       "698      chr1   1938054  rs113907178;rs35385292  ACT       A  148.60      .   \n",
       "...       ...       ...                     ...  ...     ...     ...    ...   \n",
       "1554666  chrY  56881158                       .    A      AG  129.97      .   \n",
       "1554696  chrY  56885195              rs56282167   CA       C   97.60      .   \n",
       "1554697  chrY  56885248              rs55700828    T       A  109.64      .   \n",
       "1554706  chrY  56886312                       .    G       A   89.64      .   \n",
       "1554707  chrY  56886334               rs4047361    A       C  109.64      .   \n",
       "\n",
       "                                                      INFO          FORMAT  \\\n",
       "42       AC=1;AF=0.500;AN=2;BaseQRankSum=-1.156;DP=14;E...  GT:AD:DP:GQ:PL   \n",
       "91       AC=1;AF=0.500;AN=2;BaseQRankSum=-1.440;DB;DP=1...  GT:AD:DP:GQ:PL   \n",
       "441      AC=1;AF=0.500;AN=2;BaseQRankSum=0.598;DB;DP=7;...  GT:AD:DP:GQ:PL   \n",
       "669      AC=1;AF=0.500;AN=2;BaseQRankSum=-3.087;DB;DP=1...  GT:AD:DP:GQ:PL   \n",
       "698      AC=1;AF=0.500;AN=2;BaseQRankSum=-0.180;DB;DP=8...  GT:AD:DP:GQ:PL   \n",
       "...                                                    ...             ...   \n",
       "1554666  AC=1;AF=0.500;AN=2;BaseQRankSum=-1.006;DP=13;E...  GT:AD:DP:GQ:PL   \n",
       "1554696  AC=1;AF=0.500;AN=2;BaseQRankSum=0.160;DB;DP=11...  GT:AD:DP:GQ:PL   \n",
       "1554697  AC=1;AF=0.500;AN=2;BaseQRankSum=-0.815;DB;DP=1...  GT:AD:DP:GQ:PL   \n",
       "1554706  AC=1;AF=0.500;AN=2;BaseQRankSum=1.253;DP=13;Ex...  GT:AD:DP:GQ:PL   \n",
       "1554707  AC=1;AF=0.500;AN=2;BaseQRankSum=0.256;DB;DP=14...  GT:AD:DP:GQ:PL   \n",
       "\n",
       "                          DO52567 GENOTYPE ALLELIC DEPTH DEPTH  \\\n",
       "42        0/1:6,4:10:99:150,0,219      0/1           6,4    10   \n",
       "91         0/1:3,3:6:99:117,0,117      0/1           3,3     6   \n",
       "441        0/1:3,3:6:99:112,0,112      0/1           3,3     6   \n",
       "669       0/1:8,4:12:99:144,0,324      0/1           8,4    12   \n",
       "698        0/1:4,4:8:99:156,0,156      0/1           4,4     8   \n",
       "...                           ...      ...           ...   ...   \n",
       "1554666     0/1:4,4:8:99:139,0,99      0/1           4,4     8   \n",
       "1554696   0/1:7,4:11:99:105,0,225      0/1           7,4    11   \n",
       "1554697  0/1:10,5:15:99:117,0,272      0/1          10,5    15   \n",
       "1554706    0/1:8,4:12:97:97,0,197      0/1           8,4    12   \n",
       "1554707   0/1:7,5:12:99:117,0,183      0/1           7,5    12   \n",
       "\n",
       "         GENOTYPE QUALITY PHRED-SCALED LIKELIHOODS ALLELIC FREQUENCY  \\\n",
       "42                     99                150,0,219             0.500   \n",
       "91                     99                117,0,117             0.500   \n",
       "441                    99                112,0,112             0.500   \n",
       "669                    99                144,0,324             0.500   \n",
       "698                    99                156,0,156             0.500   \n",
       "...                   ...                      ...               ...   \n",
       "1554666                99                 139,0,99             0.500   \n",
       "1554696                99                105,0,225             0.500   \n",
       "1554697                99                117,0,272             0.500   \n",
       "1554706                97                 97,0,197             0.500   \n",
       "1554707                99                117,0,183             0.500   \n",
       "\n",
       "         MINOR ALLELE COPY NUMBER  MAJOR ALLELE COPY NUMBER  \n",
       "42                              0                         2  \n",
       "91                              0                         2  \n",
       "441                             0                         2  \n",
       "669                             0                         2  \n",
       "698                             0                         2  \n",
       "...                           ...                       ...  \n",
       "1554666                         0                         2  \n",
       "1554696                         0                         2  \n",
       "1554697                         0                         2  \n",
       "1554706                         0                         2  \n",
       "1554707                         0                         2  \n",
       "\n",
       "[36927 rows x 18 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pipeline_singleSample = Pipeline([\n",
    "        # Extracting needed information and categorizing into columns\n",
    "        ('vcfDataExtraction', main.VcfDataExtractionTransformer()),\n",
    "        # Filter quality reads\n",
    "        ('filterQuality', main.FilterQualityTransformer(percentage=90)),\n",
    "        # Merging Copy Caller results with .vcf file\n",
    "        ('copyCallsMerge', main.CopyCallsMergeTransformer(cnvnator_df)),\n",
    "        # Data preparation for PyClone and PyClone-VI\n",
    "        ('pyCloneDataPreparation', main.PyCloneTransformer(samples=300)),\n",
    "        # Data preparation for SciClone\n",
    "        ('sciCloneDataPreparation', main.SciCloneTransformer(cnvnator_df)),\n",
    "        # Creation of copy number coverage file for TitanCNA\n",
    "        ('coverageFilePreparation', main.CoverageFileTransformer(cnvnator_df)),\n",
    "        # Data preparation for TitanCNA\n",
    "        ('titanCNADataPreparation', main.TitanCNATransformer()),\n",
    "        # Data preparation for FastClone\n",
    "        ('fastCloneDataPreparation', main.FastCloneTransformer()),\n",
    "        # Running PyClone-VI tool\n",
    "        ('pyCloneVI', main.PyCloneVI()),\n",
    "        # Running TitanCNA tool\n",
    "        ('titanCNA', main.TitanCNA()),\n",
    "        # Running SciClone tool\n",
    "        ('sciClone', main.SciClone()),\n",
    "        # Running FastClone tool\n",
    "        ('fastClone', main.FastClone()),\n",
    "        \n",
    "]) \n",
    "\n",
    "final_df = pipeline_singleSample.transform(main_vcf)\n",
    "\n",
    "final_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyclone2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
